donkeycar/parts/pins.py

```diff
@@ -0,0 +1,913 @@
+"""
+pins.py - high level ttl an pwm pin abstraction.
+This is designed to allow drivers that use ttl and pwm
+pins to be reusable with different underlying libaries
+and techologies.
+
+The abstract classes InputPin, OutputPin and PwmPin 
+provide an interface for starting, using and cleaning up the pins.
+The factory functions input_pin_by_id(), output_pin_by_id()
+and pwm_pin_by_id() construct pins given a string id
+that specifies the underlying pin provider and it's attributes.
+There are implementations for the Rpi.GPIO library and 
+for the PCA9685.  
+
+Pin id allows pins to be selected using a single string to
+select from different underlying providers, numbering schemes and settings.
+
+Use Rpi.GPIO library, GPIO.BOARD pin numbering scheme, pin number 13
+ pin = input_pin_by_id("RPI_GPIO.BOARD.13")
+
+Use Rpi.GPIO library, GPIO.BCM broadcom pin numbering scheme, gpio pin number 33
+ pin = output_pin_by_id("RPI_GPIO.BCM.33")
+
+Use PCA9685 on bus 0 at address 0x40, channel 7
+ pin = pwm_pin_by_id("PCA9685.0:40.7")
+
+"""
+from abc import ABC, abstractmethod
+
+from donkeycar.parts import  actuator
+
+
+class PinState:
+    LOW:int = 0
+    HIGH:int = 1
+    NOT_STARTED:int = -1
+
+
+class PinEdge:
+    RISING:int = 1
+    FALLING:int = 2
+    BOTH:int = 3
+
+
+class PinPull:
+    PULL_NONE:int = 1
+    PULL_UP:int = 2
+    PULL_DOWN:int = 3
+
+
+class PinProvider:
+    RPI_GPIO = "RPI_GPIO"
+    PCA9685 = "PCA9685"
+    PIGPIO = "PIGPIO"
+
+
+class PinScheme:
+    BOARD = "BOARD"  # board numbering
+    BCM = "BCM"      # broadcom gpio numbering
+
+
+# 
+##### Base interface for input/output/pwm pins
+##### Implementations derive from these abstact classes
+#
+
+class InputPin(ABC):
+
+    def __init__(self) -> None:
+        super().__init__()
+
+    @abstractmethod
+    def start(self, on_input=None, edge=PinEdge.RISING) -> None:
+        """
+        Start the pin in input mode.
+        on_input: function to call when an edge is detected, or None to ignore
```

Comment:
Should we use shpinx or numpy  syntax `:param on_input:` and `:return: ...` in order to get a nice render in an IDE?

Reply:
I fixed this in pins.py

Reply:
I'm still working on other files for this.

Reply:
Update actuators.py; that and pins.py are the files where I added the most code.